File: tests/property/python/areas/function_fields.py
"""Live-SageMath oracle for rational function fields over prime fields."""

from sage.all import *


def _ints(xs):
    return ','.join(str(ZZ(x)) for x in xs)


def _element(K, coeffs):
    return K(K._ring(coeffs))


def _parts(f):
    return '%s/%s' % (_ints(f.numerator().list()), _ints(f.denominator().list()))


def ff_arithmetic(p, numerator, denominator, exponent):
    K = FunctionField(GF(ZZ(p)), 'x')
    f = _element(K, numerator) / _element(K, denominator)
    g = f ** ZZ(exponent)
    return 'f=%s pow=%s degree=%s vx=%s square=%s' % (
        _parts(f),
        _parts(g),
        f.degree(),
        f.valuation(K.gen()),
        f.is_square(),
    )


def ff_factor(p, numerator, denominator):
    K = FunctionField(GF(ZZ(p)), 'x')
    f = _element(K, numerator) / _element(K, denominator)
    F = f.factor()
    factors = ';'.join('%s:%s' % (_parts(q), e) for q, e in F)
    return 'unit=%s factors=%s' % (_parts(K(F.unit())), factors)


def ff_places(p, degree):
    K = FunctionField(GF(ZZ(p)), 'x')
    places = K.places(int(degree))
    return ';'.join(str(P) for P in places)


FUNCTIONS = {
    'ff_arithmetic': ff_arithmetic,
    'ff_factor': ff_factor,
    'ff_places': ff_places,
}

File: tests/property/typescript/areas/function_fields.ts
/**
 * sagemath-ts side of the live rational-function-field differential area.
 */

import { GF } from '../../../../packages/sagemath-ts/src/rings/finite_rings/index.js';
import type {
  ConstantField,
  ConstantFieldElement,
} from '../../../../packages/sagemath-ts/src/rings/function_field/constant_field.js';
import type { FunctionFieldElement_rational } from '../../../../packages/sagemath-ts/src/rings/function_field/element_rational.js';
import type { RationalFunctionField_global } from '../../../../packages/sagemath-ts/src/rings/function_field/function_field_rational.js';
import { FunctionField } from '../../../../packages/sagemath-ts/src/rings/function_field/index.js';

type CE = ConstantFieldElement;
type FF = FunctionFieldElement_rational<CE>;
const bool = (value: boolean): string => (value ? 'True' : 'False');

function field(p: bigint): RationalFunctionField_global<CE> {
  return FunctionField(
    GF(p) as unknown as ConstantField<CE>,
    'x'
  ) as RationalFunctionField_global<CE>;
}

function element(K: RationalFunctionField_global<CE>, coeffs: bigint[]): FF {
  const k = K.constant_base_field();
  return K.__call__(coeffs.map((c) => k.__call__(c))) as FF;
}

const ints = (xs: Array<{ value?: bigint; toString(): string }>): string =>
  xs.map((x) => (typeof x.value === 'bigint' ? x.value : x.toString())).join(',');

function parts(f: FF): string {
  return `${ints(f.numerator().coeffs)}/${ints(f.denominator().coeffs)}`;
}

function ff_arithmetic(
  p: bigint,
  numerator: bigint[],
  denominator: bigint[],
  exponent: bigint
): string {
  const K = field(p);
  const f = element(K, numerator).div(element(K, denominator));
  const g = f.pow(exponent);
  return `f=${parts(f)} pow=${parts(g)} degree=${f.degree()} vx=${f.valuation(
    K.gen()
  )} square=${bool(f.is_square())}`;
}

function ff_factor(p: bigint, numerator: bigint[], denominator: bigint[]): string {
  const K = field(p);
  const f = element(K, numerator).div(element(K, denominator));
  const F = f.factor();
  const factors = F.factors.map(([q, e]) => `${parts(q)}:${e}`).join(';');
  return `unit=${parts(F.unit)} factors=${factors}`;
}

function ff_places(p: bigint, degree: bigint): string {
  return field(p).places(Number(degree)).map(String).join(';');
}

export const functions = {
  ff_arithmetic,
  ff_factor,
  ff_places,
};
